### PR TITLE
Add support for sharded models when TorchAO quantization is enabled

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
When passing a device_map, whether as a finegrained dict or simply `"auto"`, we cannot use the torchao quantization method currently.

<details>
<summary> Error </summary>

```python
Traceback (most recent call last):
  File "/home/aryan/work/diffusers/dump5.py", line 91, in <module>
    transformer = FluxTransformer2DModel.from_pretrained(
  File "/raid/aryan/nightly-venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
  File "/home/aryan/work/diffusers/src/diffusers/models/modeling_utils.py", line 920, in from_pretrained
    accelerate.load_checkpoint_and_dispatch(
  File "/raid/aryan/nightly-venv/lib/python3.10/site-packages/accelerate/big_modeling.py", line 613, in load_checkpoint_and_dispatch
    load_checkpoint_in_model(
  File "/raid/aryan/nightly-venv/lib/python3.10/site-packages/accelerate/utils/modeling.py", line 1690, in load_checkpoint_in_model
    if os.path.isfile(checkpoint):
  File "/home/aryan/.pyenv/versions/3.10.14/lib/python3.10/genericpath.py", line 30, in isfile
    st = os.stat(path)
TypeError: stat: path should be string, bytes, os.PathLike or integer, not dict
```
</details>

This is a quick workaround to throw a cleaner error message. The reason why we error out is because:
- https://github.com/huggingface/diffusers/blob/7ca64fd809f0f1a774cf31f99cb1b8ab0b16ce3f/src/diffusers/models/modeling_utils.py#L806
  Here, we merge the sharded checkpoints because hf_quantizer is not None and set `is_sharded` to False. This causes `model_file` to be a state dict instead of a string.
- https://github.com/huggingface/diffusers/blob/7ca64fd809f0f1a774cf31f99cb1b8ab0b16ce3f/src/diffusers/models/modeling_utils.py#L914
  Accelerate expects a file path when `load_checkpoint_and_dispatch` is called, but we try to pass a state dict.

Related:
- https://github.com/huggingface/diffusers/issues/10013
- https://github.com/huggingface/diffusers/pull/10009/files#r1873868579

Accelerate also provides `load_checkpoint_in_model` which might be usable here since we are working with a state dict here. Until we can figure out the best way to support this, let's raise a clean error. We can tackle in #10013 and work on refactoring too. `model_file` does not make sense as a variable name either when holding a state dict, which caused some confusions during debugging.

The missing case was found by @DN6, thanks! This was not detected by our fast/slow tests or me during testing `device_map` related changes because I was using unsharded single safetensors file for both SDXL and Flux. If the state dict is unsharded, `device_map` should work just fine whether you pass a string like `auto` or `balanced`, or if you pass a finegrained dict.

### CODE REVIEW

File: src/diffusers/models/modeling_utils.py [hunk 1]
```diff
@@ -803,6 +803,12 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                     subfolder=subfolder or "",
                 )
                 if hf_quantizer is not None:
+                    is_torchao_quantization_method = quantization_config.quant_method == QuantizationMethod.TORCHAO
```
Comment (yiyixuxu) [inline src/diffusers/models/modeling_utils.py:806]: can we consolidate with this bnb check (remove the bnb check and extend this error for any quantization method) 
https://github.com/huggingface/diffusers/blob/ac863934870556505f6035127ed39466e57b6002/src/diffusers/models/modeling_utils.py#L703 

this should not specific to any quantisation method, no? I run this test, for non-sharded checkpoint, both works for shared checkpoint, both throw same error 

```python
from diffusers import FluxPipeline, FluxTransformer2DModel, TorchAoConfig, BitsAndBytesConfig
import torch

sharded_model_id = "black-forest-labs/Flux.1-Dev"
single_model_path = "/raid/yiyi/flux_model_single"
dtype = torch.bfloat16

# create a non-sharded checkpoint
# transformer = FluxTransformer2DModel.from_pretrained(
#     model_id,
#     subfolder="transformer",
#     torch_dtype=dtype,
# )
# transformer.save_pretrained(single_model_path, max_shard_size="100GB")

torch_ao_quantization_config = TorchAoConfig("int8wo")
bnb_quantization_config = BitsAndBytesConfig(load_in_8bit=True)

print(f" testing non-sharded checkpoint")
transformer = FluxTransformer2DModel.from_pretrained(
    single_model_path,
    quantization_config=torch_ao_quantization_config,
    device_map="auto",
    torch_dtype=dtype,
)

print(f"torchao hf_device_map: {transformer.hf_device_map}")

transformer = FluxTransformer2DModel.from_pretrained(
    single_model_path, 
    quantization_config=bnb_quantization_config,
    device_map="auto",
    torch_dtype=dtype,
)
print(f"bnb hf_device_map: {transformer.hf_device_map}")


print(f" testing sharded checkpoint")
## sharded checkpoint
try:
    transformer = FluxTransformer2DModel.from_pretrained(
        sharded_model_id, 
        subfolder="transformer",
        quantization_config=torch_ao_quantization_config,
        device_map="auto",
        torch_dtype=dtype,
    )
    print(f"torchao: {transformer.hf_device_map}")
except Exception as e:
    print(f"error: {e}")

try:
    transformer = FluxTransformer2DModel.from_pretrained(
        sharded_model_id,
        subfolder="transformer",
        quantization_config=bnb_quantization_config,
        device_map="auto",
        torch_dtype=dtype,
)
    print(f"bnb hf_device_map: {transformer.hf_device_map}")
except Exception as e:
    print(f"error: {e}")
```

Comment (a-r-r-o-w) [reply src/diffusers/models/modeling_utils.py:806]: I don't think non-sharded works for both, no? non-sharded checkpoint only seems to work torchao at the moment. These are my results:

| method/shard | sharded | non-sharded |
|-|-|-|
| torchao | fails | works |
| bnb | fails | fails |

I tried with your code as well and get the following error when using BnB with unsharded on this branch:

```python
NotImplementedError: Currently, `device_map` is automatically inferred for quantized bitsandbytes models. Support for providing `device_map` as an input will be added in the future.
```

Whatever the automatic infer of `device_map` thing is, we are still unable to pass device_map manually, so I would put it in same bucket as failing. Btw, even if we don't pass device_map, no attribute `hf_device_map` is attached to the model, so we cannot access it. This seems like something we might have to fix with BnB specific PR.

```
AttributeError: 'FluxTransformer2DModel' object has no attribute 'hf_device_map'
```

Consolidating the checks together sounds good. Will update

**Comment by a-r-r-o-w** [comment]
@yiyixuxu As discussed, this PR now supports fully using both sharded/non-sharded checkpoints with the full glory of `device_map`.

```python
import torch
from diffusers import FluxPipeline, FluxTransformer2DModel, TorchAoConfig

model_id = "black-forest-labs/Flux.1-Dev"
dtype = torch.bfloat16

quantization_config = TorchAoConfig("int8wo")
transformer = FluxTransformer2DModel.from_pretrained(
    model_id,
    subfolder="transformer",
    quantization_config=quantization_config,
    torch_dtype=dtype,
    device_map="auto",
    cache_dir="/raid/.cache/huggingface"
)
if hasattr(transformer, "hf_device_map"):
    print(transformer.hf_device_map)

pipe = FluxPipeline.from_pretrained(model_id, transformer=transformer, torch_dtype=torch.bfloat16, device_map="balanced", cache_dir="/raid/.cache/huggingface")
prompt = "A cat holding a sign that says hello world"
image = pipe(prompt, num_inference_steps=30, guidance_scale=7.0, generator=torch.Generator().manual_seed(42)).images[0]
image.save("output.png")
```

<details>
<summary> logs </summary>

```bash
{'pos_embed': 0, 'time_text_embed': 0, 'context_embedder': 0, 'x_embedder': 0, 'transformer_blocks.0': 0, 'transformer_blocks.1': 0, 'transformer_blocks.2': 0, 'transformer_blocks.3': 0, 'transformer_blocks.4': 0, 'transformer_blocks.5': 0, 'transformer_blocks.6': 0, 'transformer_blocks.7': 0, 'transformer_blocks.8': 0, 'transformer_blocks.9': 0, 'transformer_blocks.10': 0, 'transformer_blocks.11': 0, 'transformer_blocks.12': 0, 'transformer_blocks.13': 1, 'transformer_blocks.14': 1, 'transformer_blocks.15': 1, 'transformer_blocks.16': 1, 'transformer_blocks.17': 1, 'transformer_blocks.18': 1, 'single_transformer_blocks.0': 1, 'single_transformer_blocks.1': 1, 'single_transformer_blocks.2': 1, 'single_transformer_blocks.3': 1, 'single_transformer_blocks.4': 1, 'single_transformer_blocks.5': 1, 'single_transformer_blocks.6': 1, 'single_transformer_blocks.7': 1, 'single_transformer_blocks.8': 1, 'single_transformer_blocks.9': 1, 'single_transformer_blocks.10': 1, 'single_transformer_blocks.11': 1, 'single_transformer_blocks.12': 1, 'single_transformer_blocks.13': 1, 'single_transformer_blocks.14': 1, 'single_transformer_blocks.15': 1, 'single_transformer_blocks.16': 1, 'single_transformer_blocks.17': 1, 'single_transformer_blocks.18': 1, 'single_transformer_blocks.19': 1, 'single_transformer_blocks.20': 1, 'single_transformer_blocks.21': 2, 'single_transformer_blocks.22': 2, 'single_transformer_blocks.23': 2, 'single_transformer_blocks.24': 2, 'single_transformer_blocks.25': 2, 'single_transformer_blocks.26': 2, 'single_transformer_blocks.27': 2, 'single_transformer_blocks.28': 2, 'single_transformer_blocks.29': 2, 'single_transformer_blocks.30': 2, 'single_transformer_blocks.31': 2, 'single_transformer_blocks.32': 2, 'single_transformer_blocks.33': 2, 'single_transformer_blocks.34': 2, 'single_transformer_blocks.35': 2, 'single_transformer_blocks.36': 2, 'single_transformer_blocks.37': 2, 'norm_out': 2, 'proj_out': 2}
Loading pipeline components...:  29%|████████████████████                                                  | 2/7 [00:02<00:05,  1.05s/it]You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████| 2/2 [00:34<00:00, 17.46s/it]
Loading pipeline components...: 100%|██████████████████████████████████████████████████████████████████████| 7/7 [00:39<00:00,  5.65s/it]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:23<00:00,  1.28it/s]
```
</details>

File: tests/quantization/torchao/test_torchao.py [hunk 2]
```diff
@@ -293,27 +294,51 @@ def test_offload(self):
...
+                    torch_dtype=torch.bfloat16,
+                    offload_folder=offload_folder,
+                )
+
+                self.assertTrue(quantized_model.hf_device_map == device_map_to_compare)
+
+                output = quantized_model(**inputs)[0]
+                output_slice = output.flatten()[-9:].detach().float().cpu().numpy()
+
+                expected_slice = np.array([0.3457, -0.0366, 0.0105, -0.2275, -0.4941, 0.4395, -0.166, -0.6641, 0.4375])
```
Comment (a-r-r-o-w) [inline tests/quantization/torchao/test_torchao.py:340]:
```suggestion
```

**Review by yiyixuxu — APPROVED** [review]
thanks for looking into this!

**Comment by yiyixuxu** [comment]
@a-r-r-o-w 
>In the code, notice that there is no call to pipe.to("cuda"). This would end up failing because pipeline_is_sequentially_offloaded is set to True, since accelerate adds a AlignDevicesHook to the transformer [here](https://github.com/huggingface/diffusers/blob/ec1c7a793f9cdcb924d302f121348d9bb5256597/src/diffusers/models/modeling_utils.py#L914).


you need to update this to not just bnb
https://github.com/huggingface/diffusers/blob/ba6fd6eb30de97370f06f5804d9cc0e10b5718b5/src/diffusers/pipelines/pipeline_utils.py#L414

**Comment by a-r-r-o-w** [comment]
@yiyixuxu Added an error with possible suggestions. Could you give this another look?

File: src/diffusers/pipelines/pipeline_utils.py [hunk 3]
```diff
@@ -420,6 +422,12 @@ def module_is_offloaded(module):
                 raise ValueError(
                     "You are trying to call `.to('cuda')` on a pipeline that has models quantized with `bitsandbytes`. Your current `accelerate` installation does not support it. Please upgrade the installation."
                 )
+            elif pipeline_has_torchao:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pipeline_utils.py:425]:
```suggestion
            elif pipeline_is_sequentially_offloaded and pipeline_has_torchao:
```

File: src/diffusers/quantizers/torchao/utils.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ..quantization_config import QuantizationMethod
+
+
+def _check_torchao_status(module) -> bool:
```
Comment (yiyixuxu) [inline src/diffusers/quantizers/torchao/utils.py:18]: did we use this?

Comment (a-r-r-o-w) [reply src/diffusers/quantizers/torchao/utils.py]: nope, removed now

**Pull request merged**